src/libraries/System.Text.Encodings.Web/src/System/Text/Encodings/Web/SpanUtility.cs

```diff
@@ -0,0 +1,254 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers.Binary;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace System.Text.Encodings.Web
+{
+    /// <summary>
+    /// Contains helpers for manipulating spans so that we can keep unsafe code out of the common path.
+    /// </summary>
+    internal static class SpanUtility
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool IsValidIndex<T>(ReadOnlySpan<T> span, int index)
+        {
+            return ((uint)index < (uint)span.Length) ? true : false;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool IsValidIndex<T>(Span<T> span, int index)
+        {
+            return ((uint)index < (uint)span.Length) ? true : false;
+        }
+
+        /// <summary>
+        /// Tries writing four bytes to the span. If success, returns true. If the span is not large
+        /// enough to hold four bytes, leaves the span unchanged and returns false.
+        /// </summary>
+        /// <remarks>
+        /// Parameters are intended to be constant values.
+        /// </remarks>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool TryWriteBytes(Span<byte> span, byte a, byte b, byte c, byte d)
+        {
+            if (span.Length >= 4)
+            {
+                uint value;
+                if (BitConverter.IsLittleEndian)
+                {
+                    value = ((uint)d << 24) | ((uint)c << 16) | ((uint)b << 8) | a;
```

Comment:
I believe writing the individual bytes is faster as
* this produces quite a lot of machine code
* cpus have a store buffer, so flushing to L1 will be done in bigger "chunks" anyway

A quick micro-benchmark (on kaby-lake) proves that:
```
| Method |     Mean |     Error |    StdDev | Ratio | RatioSD |
|------- |---------:|----------:|----------:|------:|--------:|
|      A | 1.335 ns | 0.0656 ns | 0.1022 ns |  1.00 |    0.00 |
|      B | 1.178 ns | 0.0620 ns | 0.0608 ns |  0.85 |    0.08 |
```
A...your code
B...individiual writes

Reply:
What's the asm output on your box? On my box this code produces a single `mov dword ptr [foo], CONST` instruction, which beats the performance of four `mov byte ptr [foo + i], CONST` instructions.

Reply:
> `mov dword ptr [foo], CONST`

Really a `const` if `a`, `b`, ... are arguments? 
Or just in the specific case where the method is inlined and the arguments can be evaluated as constant values?
In this case of course that's better.

For asm (note: I'm not on latest main-branch):
```assembly
; Bench.A()
; ...
       cmp       r8d,4
       jl        short M00_L02
       shl       ecx,18
       shl       r10d,10
       or        ecx,r10d
       shl       r9d,8
       or        ecx,r9d
       or        eax,ecx
       mov       [rdx],eax
       mov       eax,1
       jmp       short M00_L03
M00_L02:
       xor       eax,eax
M00_L03:
       ret

; Bench.B()
; ...
       cmp       r8d,4
       jl        short M00_L02
       mov       [rdx],al
       mov       [rdx+1],r9b
       mov       [rdx+2],r10b
       mov       [rdx+3],cl
       mov       eax,1
       jmp       short M00_L03
M00_L02:
       xor       eax,eax
M00_L03:
       ret
```

The micro-benchmark is very flaky, but `B` is always faster.
Although it's a _micro_ benchmark, that doesn't take into account that the store buffer may be full, only one store can be dispatched per cycle, etc. as it can be on real world workloads.

Reply:
@gfoidl The specific use case for this API is that all _value_ parameters are constants. That's also called out in the devdoc on the API. This causes the JIT to const-fold everything.